Support async FSMap objects in zarr.open

maxrjones opened the pull request:
Addresses https://github.com/zarr-developers/zarr-python/issues/2706

TODO:
* [ ] Add unit tests and/or doctests in docstrings
* [ ] Add docstrings and API docs for any new/modified user-facing classes and functions
* [ ] New/modified features documented in `docs/user-guide/*.rst`
* [ ] Changes documented as a new file in `changes/`
* [ ] GitHub Actions have all passed
* [ ] Test coverage is 100% (Codecov passes)

maxrjones commented on src/zarr/storage/_common.py:
```diff
@@ -312,8 +312,27 @@ async def make_store_path(
         else:
-            msg = f"Unsupported type for store_like: '{type(store_like).__name__}'"  # type: ignore[unreachable]
-            raise TypeError(msg)
+            try:  # type: ignore[unreachable]
+                import fsspec
+
+                if isinstance(store_like, fsspec.mapping.FSMap):
+                    if path:
+                        raise TypeError(
+                            "'path' was provided but is not used for FSMap store_like objects"
+                        )
+                    if storage_options:
+                        raise TypeError(
+                            "'storage_options was provided but is not used for FSMap store_like objects"
+                        )
+                    store = FsspecStore.from_mapper(store_like, read_only=_read_only)
+                else:
+                    raise (
+                        TypeError(f"Unsupported type for store_like: '{type(store_like).__name__}'")
+                    )
+            except ImportError:
+                raise (
+                    TypeError(f"Unsupported type for store_like: '{type(store_like).__name__}'")
+                ) from None
```
This is all ugly code because I don't want to assume that fsspec is installed or import it at the module level. I'll look into a better approach similar to xarray's [module_available](https://github.com/pydata/xarray/blob/2870710f543813f6de891b95397cc19e8623af92/xarray/namedarray/utils.py#L34-L60) code.

d-v-b replied:
could we pack this logic into a stand-alone function that basically narrows its input to a `FsspecStore` or errors? and maybe we could have some boolean variables in this module that represent if fsspec is present or not, e.g. something like this at the top of the file: 

```python
try:
  import fsspec
  has_fsspec = True
except ImportError:
  has_fsspec = False
```

maxrjones commented on src/zarr/storage/_fsspec.py:
```diff
@@ -29,6 +31,20 @@
 )
 
 
+def _make_async(fs: AbstractFileSystem) -> AsyncFileSystem:
+    try:
+        from fsspec.implementations.asyn_wrapper import AsyncFileSystemWrapper
+
+        fs = AsyncFileSystemWrapper(fs)
+    except ImportError as e:
+        raise ImportError(
+            f"The filesystem '{fs}' is synchronous, and the required "
+            "AsyncFileSystemWrapper is not available. Upgrade fsspec to version "
+            "2024.12.0 or later to enable this functionality."
+        ) from e
+    return fs
```
I made this a function because I think it'll be helpful for wrapping sync filesystems for both `from_url` and `from_mapper`.

d-v-b commented:
I recall there being issues with fsmap before, but I confess I don't really know what an fsmap is -- can someone explain what an fsmap is, how it differs from an fsspec filesystem, and why people would use one over the other? 

cc @martindurant   

I have a vague feeling that it could be useful to have a `Store` class that wraps generic `MutableMapping` instances, and maybe fsmaps could go there, but that requires me knowing more about the user context for fsmap.

martindurant commented:
FSMap was created specifically for the needs of Zarr, and it *could* have been essentially the same as the v2 FSStore, but was much quicker to get out and working within dask/fsspec.

FSMap is a dict-compatible interface (mutable-mapping) on top of a FS instance, which zarr worked with since forever and ignores some FS functionality like the file-like API.

To make it work with v3 might be complex, because zarr makes async calls, and the FSMap interface is blocking, even if the underlying FS is async. That means that there will be sync() within sync(), which might still work as zarr maintains its own event loop in a thread separate from fsspec's.

martindurant commented:
To be clear: this PR does *not* use the mapper, but constructs a normal store from the mapper's attributes. I support this path.

maxrjones commented:
thanks for the questions and answers, Davis and Martin!

IIUC there are three cases that we'd need to account for:
1. FSMap wraps an async instance of an async-compatible filesystem
	Solution - as implemented in this PR, extract the wrapped filesystem and use it to open an `FsspecStore`
2. FSMap wraps a non-async instance of a non-async-compatible filesystem
	Solution option 1 - extract the wrapped filesystem and wrap in `AsyncFileSystemWrapper` to open an `FsspecStore` as in https://github.com/zarr-developers/zarr-python/pull/2533
	Solution option 2 - if it's a "file" protocol, extract the wrapped filesystem and open a `LocalStore` rather than `FsspecStore`, for other protocols, wrap in `AsyncFileSystemWrapper` to open an `FsspecStore`
2. FSMap wraps a non-async instance of an async-compatibile filesystem
	Solution - this is the case I'm not sure about. Is there a way to convert from a sync instance to an async instance without needing to wrap it in `AsyncFileSystemWrapper`? @martindurant could you please offer guidance here?

martindurant commented:
>  Is there a way to convert from a sync instance to an async instance without needing to wrap it in AsyncFileSystemWrapper

The instance has all the arguments it was made with as attributes, so you can make a new instance with asynchronous=True from those.

>  if it's a "file" protocol, extract the wrapped filesystem and open a LocalStore rather than FsspecStore

Is there a reason to bother doing this?

maxrjones commented:
> if it's a "file" protocol, extract the wrapped filesystem and open a LocalStore rather than FsspecStore

Is there a reason to bother doing this?

I was thinking based on https://filesystem-spec.readthedocs.io/en/latest/async.html#limitations that LocalStore may be faster since it was designed around providing an async interface.

d-v-b commented:
IMO it's simpler if the path is always fsmap -> fsspecstore

martindurant commented:
> LocalStore may be faster since it was designed around providing an async interface.

I doubt it. The disk is not *really* async (at least with the standard syscalls python uses), so none of the async code should make any difference for local reads at all.

maxrjones commented:
I ran into https://github.com/zarr-developers/zarr-python/issues/2808 when trying to wrap sync filesystems. My preference for this PR would be to bring the codecov up to 100% and request that this be considered for review, with the plan that the conversion of sync instances to async and the wrapping of sync filesystems could be handled separately. My reasoning is that some people could use this current implementation and I'm not sure how quickly I'll be able to find time to add the other missing features.

dcherian commented:
@martindurant are you able to give this a review today? We'd like to get out a release with this bugfix asap.

maxrjones commented:
FYI I believe `zarr.open` should now work with both async and sync instances of async filesystem implementations that are created using `get_mapper` (e.g., s3, gcs). However, local filesystems seem broken with details in https://github.com/zarr-developers/zarr-python/issues/2808. TBH I'm not sure if there are other sync implementations and whether they work. This bug already exists for the `from_url` approach as far as I can tell (I'm not sure if xarray gets around this or it just wasn't caught by tests). I'm not sure how to approach the fact that the bug would be exposed more to users the new `from_mapper` function. Perhaps we should raise NotImplementedError for sync implementations in `from_wrapper` to prevent more people from running into the bug?

dcherian commented:
> Perhaps we should raise NotImplementedError for sync implementations in from_wrapper to prevent more people from running into the bug?

This sounds good to me. It's better to raise informative error messages when we can.

martindurant commented:
> how to approach the fact that the bug would be exposed more to users with the new from_mapper function

These people are currently broken, so I don't think that giving them a "doesn't work" error is any better.

maxrjones commented:
@martindurant I get some of these warnings if I only run the tests in `test/test_store/test_fsspec.py`, but not if I run the full test suite. Do you have advice on whether/how this should be avoided?

```bash
Unclosed client session
client_session: <aiohttp.client.ClientSession object at 0x11f2b5d10>
Unclosed client session
client_session: <aiohttp.client.ClientSession object at 0x118259a90>
Unclosed client session
client_session: <aiohttp.client.ClientSession object at 0x107b0ec10>
```

maxrjones commented:
> > how to approach the fact that the bug would be exposed more to users with the new from_mapper function
> 
> These people are currently broken, so I don't think that giving them a "doesn't work" error is any better.

Your advice in https://github.com/zarr-developers/zarr-python/issues/2808#issuecomment-2660034931 works - in this PR we now raise an informative error if `auto_mkdir=True` is not set in a LocalFilesystem.

martindurant commented:
I suppose that's OK, but why doesn't the store make directories? Of course, for object stores it doesn't matter...

martindurant commented:
There is one failure due to a not-captured warning. It will need to be handled to get the greens.

martindurant reviewed (commented):
My comments amount to only nitpicks.

maxrjones commented on src/zarr/storage/_fsspec.py:
```diff
@@ -30,6 +33,48 @@
+            fs_dict["asynchronous"] = True
+            return AbstractFileSystem.from_dict(fs_dict)
+        except AttributeError:
+            # Older fsspec specification used to_json rather than to_dict
```
I didn't originally notice that the older to_json/from_json functionality still exists, so I updated this section to use that in all cases in [`656a565` (#2774)](https://github.com/zarr-developers/zarr-python/pull/2774/commits/656a5656866f65d44dcc0be870d59e46957afa8b)

rossbar commented:
A quick note - I tried this PR on [the example I dropped in #2706](https://github.com/zarr-developers/zarr-python/issues/2706#issuecomment-2686222636) and got the following:

```python
# See linked example for MRE setup
>>> z = zarr.group(store=store)
Traceback (most recent call last)
   ...
AttributeError: module 'fsspec.implementations' has no attribute 'asyn_wrapper'
```

Now, there may very well be a problem with my example rather than this PR - all I can say is that the example does work with zarr v2.

maxrjones commented:
> A quick note - I tried this PR on [the example I dropped in #2706](https://github.com/zarr-developers/zarr-python/issues/2706#issuecomment-2686222636) and got the following:
> 
> ```python
> # See linked example for MRE setup
> >>> z = zarr.group(store=store)
> Traceback (most recent call last)
>    ...
> AttributeError: module 'fsspec.implementations' has no attribute 'asyn_wrapper'
> ```
> 
> Now, there may very well be a problem with my example rather than this PR - all I can say is that the example does work with zarr v2.

Thanks for the feedback! Could you please which version of fsspec you're using?

maxrjones commented on src/zarr/storage/_common.py:
```diff
@@ -311,9 +317,18 @@ async def make_store_path(
         else:
-            msg = f"Unsupported type for store_like: '{type(store_like).__name__}'"  # type: ignore[unreachable]
-            raise TypeError(msg)
+            raise (TypeError(f"Unsupported type for store_like: '{type(store_like).__name__}'"))
```
I'll check on the linter issue, as for the bad coverage I think we still need this in case fsspec isn't installed. https://github.com/zarr-developers/zarr-python/pull/2872 should fix the code coverage gap, but it's taking longer than I expected to finish that because it uncovered other issues with the test matrix design (working on it but might not finish today)

maxrjones replied:
thanks, I applied these suggestions in [`466bdd2` (#2774)](https://github.com/zarr-developers/zarr-python/pull/2774/commits/466bdd2e5ade946aa539e4351abede4e37fde525)

rossbar commented:
> Thanks for the feedback! Could you please which version of fsspec you're using?

The fsspec version is `2025.2.0`: full output of `pip list` below!

<details>
  <summary>env</summary>
<pre>
Package                 Version
----------------------- ---------------------
asttokens               3.0.0
asyncssh                2.20.0
cffi                    1.17.1
crc32c                  2.7.1
cryptography            44.0.1
decorator               5.2.1
Deprecated              1.2.18
donfig                  0.8.1.post1
executing               2.2.0
fsspec                  2025.2.0
ipython                 9.0.0
ipython_pygments_lexers 1.1.1
jedi                    0.19.2
matplotlib-inline       0.1.7
numcodecs               0.15.1
numpy                   2.2.3
packaging               24.2
parso                   0.8.4
pexpect                 4.9.0
pip                     25.0
prompt_toolkit          3.0.50
ptyprocess              0.7.0
pure_eval               0.2.3
pycparser               2.22
Pygments                2.19.1
PyYAML                  6.0.2
sshfs                   2025.2.0
stack-data              0.6.3
traitlets               5.14.3
typing_extensions       4.12.2
wcwidth                 0.2.13
wrapt                   1.17.2
zarr                    3.0.5.dev24+gd4d22560
</pre>
</details>

maxrjones commented:
> How close is this PR to getting merged? Presently, this issue is blocking our ability use zarr-python v3 out of the box with pygeoapi for a USGS deployment. See [code.usgs.gov/wma/nhgf/pygeoapi/-/issues/56](https://code.usgs.gov/wma/nhgf/pygeoapi/-/issues/56)
> 
> We could of course apply a fix to pygeoapi directly, using suggestions from:
> 
> * [zarr v3 does not accept stores of type FSMap #2706](https://github.com/zarr-developers/zarr-python/issues/2706)
> 
> ... but this PR could make things much easier for us and others!
> 
> cc: @kieranbartels

Thanks for the nudge on this PR. I totally get that it's a lot to expect all users to implement a workaround given how widely used the pattern was for Zarr-Python 2.

I'll spend a couple hours tomorrow morning cleaning this up before the Zarr Python developers meeting and should be able to get back to you with a timeline then.

maxrjones commented on src/zarr/storage/_common.py:
```diff
@@ -228,7 +234,7 @@ def __eq__(self, other: object) -> bool:
 
 async def make_store_path(
-    store_like: StoreLike | None,
+    store_like: StoreLike | FSMap | None,
```
`from __future__ import annotations` makes it so you don't need the `else FSMap = None` block

maxrjones commented:
I think this is ready for a final review, as I addressed all the comments and increased the test coverage. Would you have time to try for a review before the next release @d-v-b @martindurant? Thanks for your support with the PR

d-v-b commented on src/zarr/storage/_common.py:
```diff
@@ -314,9 +319,16 @@ async def make_store_path(
             store = await MemoryStore.open(store_dict=store_like, read_only=_read_only)
+        elif _has_fsspec and isinstance(store_like, FSMap):
+            if path:
+                raise ValueError("'path' was provided but is not used for FSMap store_like objects")
```
Can we add a constructive instruction to this error? Something like "...set path to `None`, or leave it unset, when specifying a store with an FSMap"

maxrjones commented on src/zarr/storage/_fsspec.py:
```diff
@@ -137,6 +178,39 @@ def from_upath(
+            Whether the store is read-only, defaults to False.
+        allowed_exceptions : tuple, optional
+            The exceptions that are allowed to be raised when accessing the
+            store. Defaults to ALLOWED_EXCEPTIONS.
```
Making it an f-string leads to the linter error "B021 f-string used as docstring. Python will interpret this as a joined string, rather than a docstring.". Since the current usage is consistent with other docstrings in the codebase, I'd prefer to leave this as is for now

maxrjones commented on src/zarr/storage/_common.py:
```diff
@@ -314,9 +319,16 @@ async def make_store_path(
+                raise ValueError("'path' was provided but is not used for FSMap store_like objects")
+            if storage_options:
+                raise ValueError(
+                    "'storage_options was provided but is not used for FSMap store_like objects"
```
thanks, addressed in [`7f48751` (#2774)](https://github.com/zarr-developers/zarr-python/pull/2774/commits/7f4875122928b3cd2e834d3d4d005e7661550a27)

maxrjones commented:
@d-v-b what was your fix the last time you ran into the upstream fsspec <-> s3fs incompatibility issue?

d-v-b commented:
note that the problem went away without me doing anything, presumably due to some activity over in fsspec?

maxrjones commented:
@martindurant thanks for your earlier review of this PR. I just wanted to check if you want to take another look before it get merged?

d-v-b commented on src/zarr/storage/_fsspec.py:
```diff
@@ -30,6 +32,45 @@
+    If the filesystem class does not support async operations, the existing instance
+    is wrapped with AsyncFileSystemWrapper.
+    """
+    import fsspec
+    from packaging.version import parse as parse_version
```
why do we import these here?

maxrjones replied:
fsspec is imported here to follow the conventional of importing optional dependencies in the functions in which they are used, based on conversations from the obstore integration PR.

parse can be imported in the module scope so I'll change that.

maxrjones replied:
I moved the packaging.version.parse import in [`3b06c53` (#2774)](https://github.com/zarr-developers/zarr-python/pull/2774/commits/3b06c536cc5398a83b81eea7fc95abcdf3eaa37f)

d-v-b commented on src/zarr/storage/_fsspec.py:
```diff
@@ -30,6 +32,45 @@
+        return fs
+    if fs.async_impl:
+        # Convert sync instance of an async fs to an async instance
+        import json
```
can we declare this import at the top of the file?

maxrjones replied:
moved in [`3b06c53` (#2774)](https://github.com/zarr-developers/zarr-python/pull/2774/commits/3b06c536cc5398a83b81eea7fc95abcdf3eaa37f)

d-v-b commented on src/zarr/storage/_fsspec.py:
```diff
@@ -137,6 +178,39 @@ def from_upath(
+        FsspecStore
+        """
+        if not fs_map.fs.async_impl or not fs_map.fs.asynchronous:
+            fs_map.fs = _make_async(fs_map.fs)
```
can we avoid mutating the `fs_map` object here? It looks like we only need the `fs` and `root` attributes, so lets pull those off into their own variables.

d-v-b reviewed (changes requested):
some minor questions about imports and re-assignment of object attributes, but otherwise this looks good

rabernat commented:
@aufdenkampe - can you explain how you are using FSMap today? There may be much simpler workarounds than the solution in this PR. For example, are you loading data from object storage? Do you know the bucket and path name of the store you are trying to open?
